service/src/main/kotlin/io/provenance/api/domain/usecase/provenance/tx/scope/ChangeScopeOwnership.kt

```diff
@@ -0,0 +1,81 @@
+import io.provenance.scope.util.ProtoJsonUtil.toJson
+import org.springframework.stereotype.Component
+
+@Component
+class ChangeScopeOwnership(
+    private val provenance: Provenance,
+    private val entityManager: EntityManager,
+    private val getSigner: GetSigner,
+    private val provenanceProperties: ProvenanceProperties,
+) : AbstractUseCase<ChangeScopeOwnershipRequestWrapper, TxResponse>() {
+    override suspend fun execute(args: ChangeScopeOwnershipRequestWrapper): TxResponse {
+        require(args.request.newValueOwner !== null || args.request.newDataAccess !== null) {
+            "Must request at least one change to the scope"
+        }
+
+        val signer = getSigner.execute(GetSignerRequest(args.uuid, args.request.account))
+
+        val scopeResponse = provenance.getScope(args.request.provenanceConfig, args.request.scopeId)
+            .takeIf { response -> response.scope.isSet() }
+            ?: throw NotFoundError("No scope found")
+
+        scopeResponse.scope.scopeIdInfo.scopeUuid.let { actualScopeId ->
+            require(actualScopeId === args.request.scopeId.toString()) {
+                "Expected to fetch a scope with UUID ${args.request.scopeId} but got a UUID of $actualScopeId"
+            }
+        }
+
+        val message = MsgWriteScopeRequest.newBuilder().apply {
+            scopeUuid = args.request.scopeId.toString()
+            specUuid = scopeResponse.scope.scopeSpecIdInfo.scopeSpecUuid
+            scope = scopeResponse.scope.scope.toBuilder().also { scopeBuilder ->
+                args.request.newValueOwner?.let { requestedNewValueOwner ->
+                    /** Only change value owner if [valueOwnerAddress][args.request.valueOwnerAddress] was not null. */
+                    scopeBuilder.valueOwnerAddress = requestedNewValueOwner
+                }
+                /** Only change data access if [newDataAccess][args.request.newDataAccess] was not null. */
+                args.request.newDataAccess?.let { requestedNewDataAccess ->
+                    entityManager.hydrateKeys(requestedNewDataAccess).toMessageSet(
+                        isMainnet = provenanceProperties.mainnet
+                    ).let { additionalAudiences ->
+                        scopeBuilder.clearDataAccess()
+                        scopeBuilder.addAllDataAccess(additionalAudiences)
+                    }
+                }
+            }.build()
+            addSigners(signer.address())
+        }.build()
+
+        return provenance.onboard(
+            args.request.provenanceConfig.chainId,
+            args.request.provenanceConfig.nodeEndpoint,
+            signer,
+            listOf(message.toAny()).toTxBody().let { cosmosTxBody ->
+                TxBody(
+                    json = ObjectMapper().readValue(cosmosTxBody.toJson(), ObjectNode::class.java),
+                    base64 = cosmosTxBody.messagesList.map { cosmosTxBody.toByteArray().toBase64String() },
+                )
```

Comment:
I tried extracting this conversion — from the Cosmos `TxBody` to the model — to an extension function, but it would throw an `InvalidProtocolBufferException` in the `CreateTxTest` that it can't find the type URL for the `MsgScopeWriteRequest`.

Reply:
Yes, just like that. Will look into it later I guess.

Reply:
Fixed — the cause was usage of
```kotlin
messagesList.map { toByteArray().toBase64String() }
```
instead of
```kotlin
messagesList.map { it.toByteArray().toBase64String() }
```
I must have misplaced the `it` while moving the function around and then spent way too long trying to figure out why the base64 decoding was failing.